src/impl/dtlstransport.cpp

```diff
@@ -360,7 +360,308 @@ int DtlsTransport::TimeoutCallback(gnutls_transport_ptr_t ptr, unsigned int /* m
+
+	if (dtlsTransport->fin_ms == 0) {
+		return -1;
+	} else if (elapsed >= dtlsTransport->fin_ms) {
+		return 2;
+	} else if (elapsed >= dtlsTransport->int_ms) {
+		return 1;
+	} else {
+		return 0;
+	}
+}
+
+
+bool DtlsTransport::send(message_ptr message) {
+	if (!message || state() != State::Connected)
+		return false;
+
+	PLOG_VERBOSE << "Send size=" << message->size();
+
+	int ret;
+	do {
+		std::lock_guard lock(mSendMutex);
+		mCurrentDscp = message->dscp;
+
+		if (message->size() > size_t(mbedtls_ssl_get_max_out_record_payload(&mSsl)))
+			return false;
+
+		auto buff = &(message->data())[0];
+
+		ret = mbedtls_ssl_write(&mSsl, (const unsigned char *) buff, message->size());
+	} while (ret == MBEDTLS_ERR_SSL_WANT_WRITE);
+
+	mbedtls::check(ret);
+
+	return mOutgoingResult;
+}
+
+void DtlsTransport::incoming(message_ptr message) {
+	if (!message) {
+		mIncomingQueue.stop();
+		return;
+	}
+
+	PLOG_VERBOSE << "Incoming size=" << message->size();
+	mIncomingQueue.push(message);
+	enqueueRecv();
+}
+
+
+bool DtlsTransport::outgoing(message_ptr message) {
+	message->dscp = mCurrentDscp;
+
+	bool result = Transport::outgoing(std::move(message));
+	mOutgoingResult = result;
+	return result;
+}
+
+bool DtlsTransport::demuxMessage(message_ptr) {
+	// Dummy
+	return false;
+}
+
+void DtlsTransport::postHandshake() {
+	// Dummy
+}
+
+void DtlsTransport::doRecv() {
+	std::lock_guard lock(mRecvMutex);
+	--mPendingRecvCount;
+
+	if (state() != State::Connecting && state() != State::Connected)
+		return;
+
+	try {
+		const size_t bufferSize = 4096;
+		char buffer[bufferSize];
+
+		// Handle handshake if connecting
+		if (state() == State::Connecting) {
+			auto handshakeStatus = mbedtls_ssl_handshake(&mSsl);
+			switch (handshakeStatus) {
+				case 0:
+					break;
+				case MBEDTLS_ERR_SSL_WANT_WRITE:
+				case MBEDTLS_ERR_SSL_WANT_READ:
+				case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+				case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS: {
+					auto timeout = milliseconds(this->fin_ms);
+					ThreadPool::Instance().schedule(timeout, [weak_this = weak_from_this()]() {
+						if (auto locked = weak_this.lock())
+							locked->doRecv();
+					});
+					return;
+				}
+				default:
+				mbedtls::check(handshakeStatus);
+			}
+
+			PLOG_INFO << "DTLS handshake finished";
+			changeState(State::Connected);
+			postHandshake();
+		}
+
+		if (state() == State::Connected) {
+			while (true) {
+				auto ret = mbedtls_ssl_read(&mSsl, (unsigned char *) buffer, bufferSize);
+				if (ret == 0 || ret == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+					// Closed
+					PLOG_DEBUG << "DTLS connection cleanly closed";
+					break;
+				}
+				switch (ret) {
+					case MBEDTLS_ERR_SSL_WANT_READ:
+					case MBEDTLS_ERR_SSL_WANT_WRITE:
+					case MBEDTLS_ERR_SSL_ASYNC_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CRYPTO_IN_PROGRESS:
+					case MBEDTLS_ERR_SSL_CLIENT_RECONNECT:
+						return;
+				}
```

Comment:
I would use a simple if here for readability. Also, do all of those return codes make sense here? For instance, shouldn't the connection be closed instead on client reconnect?

Reply:
`MBEDTLS_SSL_RENEGOTIATION_DISABLED` is the default, so we don't need to handle it ourselves. Converted the other statements to an if.